src/github.com/couchbase/sync_gateway/rest/handler.go

```diff
@@ -156,8 +170,26 @@ func (h *handler) invoke(method handlerMethod) error {
 		if err != nil {
 			return err
 		}
+
+		//If the Database is currently online process as normal
+		dbState := atomic.LoadUint32(&dbContext.State)
+		if (dbState == db.DBOnline) {
+			//Take a readlock for any method that does not run offline, run offline methods will take a write lock
```

Comment:
I have removed this generic statement, it is the _offline code that takes the write lock to block on existing open calls